crates/types/src/replication/balanced_spread_selector.rs

```diff
@@ -0,0 +1,979 @@
+            &mut heap,
+            &mut selected,
+        )?;
+    }
+    if let Some(required_zones) = replication_property.copies_at_scope(LocationScope::Zone) {
+        populate_scope(
+            LocationScope::Zone,
+            required_zones,
+            &mut heap,
+            &mut selected,
+        )?;
+    }
+
+    // # Phase 2 – Balanced completion with viability guard
+    let total_num_copies = replication_property.num_copies();
+    while selected.len() < total_num_copies as usize {
+        let mut bucket = heap.pop().ok_or(Error::InsufficientNodes {
+            scope: LocationScope::Node,
+            available: selected.len(),
+            replication_factor: total_num_copies,
+        })?;
+        if let Some(candidate) = bucket.next_candidate() {
+            if is_candidate_viable(&selected, &candidate, replication_property) {
+                selected.insert(candidate);
+            } else {
+                // Candidate would dead‑end us – skipping it.
+            }
+        }
+        if !bucket.is_empty() {
+            heap.push(bucket);
+        }
+    }
+
+    Ok(selected.nodes)
+}
+
+/// Tracks the domains already represented in the output nodeset.
+struct SelectionState {
+    nodes: NodeSet,
+    regions: HashSet<RegionKey>, // indexes regions inside Topology::regions
+    zones: HashSet<(RegionKey, ZoneKey)>, // indexes inside Topology
+}
+
+impl SelectionState {
+    fn new(replication_property: &ReplicationProperty) -> Self {
+        Self {
+            nodes: NodeSet::with_capacity(replication_property.num_copies() as usize),
+            regions: replication_property
+                .copies_at_scope(LocationScope::Region)
+                .map(|r| HashSet::with_capacity(r as usize))
+                .unwrap_or_default(),
+            zones: replication_property
+                .copies_at_scope(LocationScope::Zone)
+                .map(|z| HashSet::with_capacity(z as usize))
+                .unwrap_or_default(),
+        }
+    }
+
+    fn insert(&mut self, node: Node) {
+        self.nodes.insert(node.id);
+        if let Some(r) = node.region_idx {
+            self.regions.insert(r);
+        }
+        if let (Some(r), Some(z)) = (node.region_idx, node.zone_idx) {
+            self.zones.insert((r, z));
+        }
+    }
+
+    fn len(&self) -> usize {
+        self.nodes.len()
+    }
+
+    fn num_distinct_regions(&self) -> usize {
+        self.regions.len()
+    }
+    fn num_distinct_zones(&self) -> usize {
+        self.zones.len()
+    }
+
+    fn has_region(&self, r: RegionKey) -> bool {
+        self.regions.contains(&r)
+    }
+    fn has_zone(&self, rz: (RegionKey, ZoneKey)) -> bool {
+        self.zones.contains(&rz)
+    }
+}
+
+#[derive(Debug)]
+struct CandidateNode {
+    node: Node,
+    hash: u64,
+    preferred: bool,
+}
+
+impl Ord for CandidateNode {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        match (self.preferred, other.preferred) {
```

Comment:
Preferences only work within a `DomainBucket`, right? If I have replication property `{region: 1}` and a preferred node 1 and the available set of nodes is:

```
1: region-1
2: region-2
```

it can happen that the spread selector chooses `2` (depending on the region hash). Is this a deliberate choice atm?